### ISSUE TITLE
自己写的动态图 CRNN 模型训练一直不正确

### ISSUE THREAD

**Issue by xiangyubo** [PaddlePaddle/Paddle#22906]
1）PaddlePaddle版本：1.6.2
    3）GPU：GPU v100
    4）系统环境：Python3.7，AI Studio 在线 notebook 环境
- 训练信息
    1）单机单卡
    2）显存信息，16G显存，32G内存

我自己照着静态图的 CRNN 和一些动态图的模型，自己写了一个动态图版本的 CRNN 模型。流程能跑通。训练时候 loss 看着很正常，第一个 batch 的 loss 是40+，运行20个 batch 之后下降到7-8，之后继续下降到4-5。但是解码出来的东西一直不对，字段级别的准确率一直是0。
解码部分我一开始用的是 fluid.layers.ctc_greedy_decoder，结果不对。我觉得这个太复杂，自己写了一个类似的，但依然不对。求帮忙看看是哪个部分不对，导致准确率一直为0。

AI Studio 环境地址：https://aistudio.baidu.com/aistudio/projectdetail/251819
里面带着数据集和完整代码，用GPU方式运行，进去解压数据直接训练就可以。项目已经公开

**Comment by LDOUBLEV** [comment]
请问，你参考的是哪个CRNN静态图实现，训练的Loss很大，也不太正常，是不是没有训练好？

**Comment by xiangyubo** [comment]
> 请问，你参考的是哪个CRNN静态图实现，训练的Loss很大，也不太正常，是不是没有训练好？

不是没训练好，我训练了50轮还是这个情况。参考的就是 models 里的静态图

**Comment by LDOUBLEV** [comment]
问题定位到warp_ctc这个API，在动态图中使用，计算loss出现inf，后续找API负责人check下是不是API的问题，有消息在comment 里回复。

动态图里使用warp_ctc的代码。
```bash 
import paddle.fluid as fluid
import numpy as np
batch_size = 128
max_seq_length = 32
num_classes = 64
with fluid.dygraph.guard(fluid.CPUPlace()):
    x = fluid.dygraph.to_variable(np.random.rand(max_seq_length, batch_size, num_classes).astype('float32'))
    y = fluid.dygraph.to_variable(np.random.randint(0, num_classes+1, [batch_size, max_seq_length]).astype('int32'))
    x_len = fluid.dygraph.to_variable(np.array([max_seq_length] * batch_size).astype('int64'))
    y_len = fluid.dygraph.to_variable(np.array([max_seq_length] * batch_size).astype('int64'))
    cost = fluid.layers.warpctc(input=x, label=y, blank=0, input_length=x_len, label_length=y_len)
    avg_loss = fluid.layers.reduce_mean(cost)
    print(avg_loss.numpy())
```
输出结果是inf

![image](https://user-images.githubusercontent.com/26592129/76159770-a6b63300-615e-11ea-9534-750bb62a7a05.png)

**Comment by xiangyubo** [comment]
> 问题定位到warp_ctc这个API，在动态图中使用，计算loss出现inf，后续找API负责人check下是不是API的问题，有消息在comment 里回复。
> 
> 动态图里使用warp_ctc的代码。
> 
> ```shell
> import paddle.fluid as fluid
> import numpy as np
> batch_size = 128
> max_seq_length = 32
> num_classes = 64
> with fluid.dygraph.guard(fluid.CPUPlace()):
>     x = fluid.dygraph.to_variable(np.random.rand(max_seq_length, batch_size, num_classes).astype('float32'))
>     y = fluid.dygraph.to_variable(np.random.randint(0, num_classes+1, [batch_size, max_seq_length]).astype('int32'))
>     x_len = fluid.dygraph.to_variable(np.array([max_seq_length] * batch_size).astype('int64'))
>     y_len = fluid.dygraph.to_variable(np.array([max_seq_length] * batch_size).astype('int64'))
>     cost = fluid.layers.warpctc(input=x, label=y, blank=0, input_length=x_len, label_length=y_len)
>     avg_loss = fluid.layers.reduce_mean(cost)
>     print(avg_loss.numpy())
> ```
> 
> 输出结果是inf
> 
> ![image](https://user-images.githubusercontent.com/26592129/76159770-a6b63300-615e-11ea-9534-750bb62a7a05.png)

好的，谢谢了

**Comment by lfchener** [comment]
如下图，通常在 ``warp-ctc`` 计算的过程中，``label`` 的 ``len`` 要比 ``x`` 的小很多，所以 ``y_len`` 要比 ``max_seq_length`` 小，改小红框中的参数后，计算不会出现 inf 的情况。
![image](https://user-images.githubusercontent.com/6771821/76218572-684c7100-624f-11ea-9cd4-21b963015980.png)

而在静态图中，如果使用 ``y_len = max_seq_length`` 的配置，也会报inf，所以排除 ``warp-ctc``在动态图中的问题。建议 @xiangyubo 和静态图对齐一下实现，可能是实现的 diff 导致的无法达到训练效果
